In [1]:
import pandas as pd

In [2]:
odds =pd.read_parquet("data/processed/odds.parquet")
point_by_point =pd.read_parquet("data/processed/point_by_point.parquet")
statistics =pd.read_parquet("data/processed/statistics.parquet")
tennis_power =pd.read_parquet("data/processed/tennis_power.parquet")
votes =pd.read_parquet("data/processed/votes.parquet")
MatchAwayScoreInfo=pd.read_parquet("data/processed/match_away_score_info.parquet")
MatchHomeScoreInfo=pd.read_parquet("data/processed/match_home_score_info.parquet")
MatchEventInfo=pd.read_parquet("data/processed/match_event_info.parquet")
MatchAwayTeamInfo=pd.read_parquet("data/processed/MatchAwayTeamInfo.parquet")
MatchHomeTeamInfo=pd.read_parquet("data/processed/MatchHomeTeamInfo.parquet")
MatchRoundInfo=pd.read_parquet("data/processed/MatchRoundInfo.parquet")
MatchSeasonInfo=pd.read_parquet("data/processed/MatchSeasonInfo.parquet")
MatchTimeInfo=pd.read_parquet("data/processed/MatchTimeInfo.parquet")
MatchTournamentInfo=pd.read_parquet("data/processed/MatchTournamentInfo.parquet")
MatchVenueInfo=pd.read_parquet("data/processed/MatchVenueInfo.parquet")

### 1. Data Structure

In [3]:
datasets={
    "MatchAwayScoreInfo":MatchAwayScoreInfo,
    "MatchHomeScoreInfo":MatchHomeScoreInfo,
    "MatchEventInfo":MatchEventInfo,
    "MatchAwayTeamInfo":MatchAwayTeamInfo,
    "MatchHomeTeamInfo":MatchHomeTeamInfo,
    "MatchRoundInfo":MatchRoundInfo,
    "MatchSeasonInfo":MatchSeasonInfo,
    "MatchTimeInfo":MatchTimeInfo,
    "MatchTournamentInfo":MatchTournamentInfo,
    "MatchVenueInfo":MatchVenueInfo,
    "odds":odds,
    "point_by_point":point_by_point,
    "statistics":statistics,
    "tennis_power":tennis_power,
    "votes":votes
}
for name,df in datasets.items():
    print(name,type(df))

MatchAwayScoreInfo <class 'pandas.core.frame.DataFrame'>
MatchHomeScoreInfo <class 'pandas.core.frame.DataFrame'>
MatchEventInfo <class 'pandas.core.frame.DataFrame'>
MatchAwayTeamInfo <class 'pandas.core.frame.DataFrame'>
MatchHomeTeamInfo <class 'pandas.core.frame.DataFrame'>
MatchRoundInfo <class 'pandas.core.frame.DataFrame'>
MatchSeasonInfo <class 'pandas.core.frame.DataFrame'>
MatchTimeInfo <class 'pandas.core.frame.DataFrame'>
MatchTournamentInfo <class 'pandas.core.frame.DataFrame'>
MatchVenueInfo <class 'pandas.core.frame.DataFrame'>
odds <class 'pandas.core.frame.DataFrame'>
point_by_point <class 'pandas.core.frame.DataFrame'>
statistics <class 'pandas.core.frame.DataFrame'>
tennis_power <class 'pandas.core.frame.DataFrame'>
votes <class 'pandas.core.frame.DataFrame'>


In [4]:
for name,df in datasets.items():
    print(name)
    print("Row:",df.shape[0])
    print("Columns:",df.shape[1])
    print("-----------------------")

MatchAwayScoreInfo
Row: 35053
Columns: 15
-----------------------
MatchHomeScoreInfo
Row: 35164
Columns: 15
-----------------------
MatchEventInfo
Row: 35053
Columns: 11
-----------------------
MatchAwayTeamInfo
Row: 24203
Columns: 19
-----------------------
MatchHomeTeamInfo
Row: 25610
Columns: 19
-----------------------
MatchRoundInfo
Row: 19283
Columns: 6
-----------------------
MatchSeasonInfo
Row: 35671
Columns: 5
-----------------------
MatchTimeInfo
Row: 35671
Columns: 8
-----------------------
MatchTournamentInfo
Row: 35671
Columns: 17
-----------------------
MatchVenueInfo
Row: 35423
Columns: 6
-----------------------
odds
Row: 60946
Columns: 12
-----------------------
point_by_point
Row: 2549369
Columns: 14
-----------------------
statistics
Row: 1358234
Columns: 14
-----------------------
tennis_power
Row: 469677
Columns: 6
-----------------------
votes
Row: 35658
Columns: 4
-----------------------


In [5]:
odds.columns

Index(['match_id', 'market_id', 'market_name', 'is_live', 'suspended',
       'initial_fractional_value', 'fractional_value', 'choice_name',
       'choice_source_id', 'winnig', 'change', 'date'],
      dtype='object')

In [6]:
statistics.columns

Index(['match_id', 'period', 'statistic_category_name', 'statistic_name',
       'home_stat', 'away_stat', 'compare_code', 'statistic_type',
       'value_type', 'home_value', 'away_value', 'home_total', 'away_total',
       'date'],
      dtype='object')

In [7]:
point_by_point.columns

Index(['match_id', 'set_id', 'game_id', 'point_id', 'home_point', 'away_point',
       'point_description', 'home_point_type', 'away_point_type', 'home_score',
       'away_score', 'serving', 'scoring', 'date'],
      dtype='object')

In [8]:
MatchAwayScoreInfo.head()


,match_id,current_score,display_score,period_1,period_2,period_3,period_4,period_5,period_1_tie_break,period_2_tie_break,period_3_tie_break,period_4_tie_break,period_5_tie_break,normal_time,date
0,11974053,2.0,2.0,NaN,NaN,NaN,None,None,NaN,NaN,NaN,None,None,None,20240201
1,11974066,4.0,4.0,NaN,NaN,NaN,None,None,NaN,NaN,NaN,None,None,None,20240201
2,11998445,2.0,2.0,7.0,2.0,7.0,None,None,NaN,NaN,7.0,None,None,None,20240201
3,11998446,2.0,2.0,6.0,6.0,NaN,None,None,NaN,NaN,NaN,None,None,None,20240201
4,11998447,0.0,0.0,4.0,0.0,NaN,None,None,NaN,NaN,NaN,None,None,None,20240201


In [9]:
MatchHomeScoreInfo.head()

,match_id,current_score,display_score,period_1,period_2,period_3,period_4,period_5,period_1_tie_break,period_2_tie_break,period_3_tie_break,period_4_tie_break,period_5_tie_break,normal_time,date
0,11974053,3.0,3.0,NaN,NaN,NaN,None,None,NaN,NaN,NaN,None,None,None,20240201
1,11974066,0.0,0.0,NaN,NaN,NaN,None,None,NaN,NaN,NaN,None,None,None,20240201
2,11998445,1.0,1.0,5.0,6.0,6.0,None,None,NaN,NaN,4.0,None,None,None,20240201
3,11998446,0.0,0.0,4.0,1.0,NaN,None,None,NaN,NaN,NaN,None,None,None,20240201
4,11998447,2.0,2.0,6.0,6.0,NaN,None,None,NaN,NaN,NaN,None,None,None,20240201


In [10]:
MatchAwayTeamInfo.head()

,match_id,name,slug,gender,user_count,residence,birthplace,height,weight,plays,turned_pro,current_prize,total_prize,player_id,current_rank,name_code,country,full_name,date
0,11998445,Auger-Aliassime F.,auger-aliassime-felix,M,23318,"Monte Carlo, Monaco","Montreal, Canada",1.93,87.0,right-handed,2017,218538.0,10166964.0,192013,30.0,AUG,Canada,"Auger-Aliassime, Felix",20240201
1,11998446,Cobolli F.,flavio-cobolli,M,5995,Rome,"Florence, Italy",1.83,71.0,right-handed,None,178358.0,808536.0,273680,69.0,COB,Italy,Flavio Cobolli,20240201
2,11998447,Martínez P.,martinez-pedro,M,2621,"Valencia, Spain","Alzira, Spain",1.85,76.0,right-handed,2016,39448.0,2732689.0,77223,101.0,MAR,Spain,"Martinez, Pedro",20240201
3,11998448,Muller A.,muller-alexandre,M,2443,France,"Poissy, France",1.83,73.0,right-handed,2014,106522.0,1178035.0,88992,81.0,MUL,France,"Muller, Alexandre",20240201
4,11998449,Mayot H.,mayot-harold,M,1492,"Marly, France","Metz, France",1.78,78.0,right-handed,None,65798.0,424547.0,248846,132.0,MAY,France,"Mayot, Harold",20240201


In [11]:
MatchHomeTeamInfo.head()

,match_id,name,slug,gender,user_count,residence,birthplace,height,weight,plays,turned_pro,current_prize,total_prize,player_id,current_rank,name_code,country,full_name,date
0,11998445,Cazaux A.,cazaux-arthur,M,4481,None,"Montpellier, France",1.83,NaN,right-handed,None,233290.0,643582.0,287803,86.0,CAZ,France,"Cazaux, Arthur",20240201
1,11998446,Lestienne C.,lestienne-constant,M,2032,"Paris, France","Amiens, France",1.80,72.0,right-handed,2012,105335.0,1404567.0,62790,99.0,LES,France,"Lestienne, Constant",20240201
2,11998447,Ćorić B.,coric-borna,M,12414,"Dubai, UAE","Zagreb, Croatia",1.88,85.0,right-handed,2013,208593.0,9947986.0,64580,31.0,COR,Croatia,"Ćorić, Borna",20240201
3,11998448,Mmoh M.,mmoh-michael,M,2206,"Bradenton, FL, USA","Riyadh, Saudi Arabia",1.88,90.0,right-handed,2016,47928.0,1775118.0,131442,122.0,MMO,USA,"Mmoh, Michael",20240201
4,11998449,Paire B.,paire-benoit,M,6089,"Geneva, Switzerland","Avignon, France",1.96,80.0,right-handed,2007,33248.0,8622687.0,22218,110.0,PAI,France,"Paire, Benoit",20240201


In [12]:
MatchEventInfo.head()

,match_id,first_to_serve,home_team_seed,away_team_seed,custom_id,winner_code,default_period_count,start_datetime,match_slug,final_result_only,date
0,11974053,NaN,None,None,NNfsQTf,1.0,3,1706878800,switzerland-netherlands,False,20240201
1,11974066,NaN,None,None,hTfsrpg,2.0,3,1706871600,ukraine-usa,False,20240201
2,11998445,1.0,None,3,nPBbsdgpc,2.0,3,1706810400,cazaux-auger-aliassime,False,20240201
3,11998446,2.0,None,None,PfAsFyjc,2.0,3,1706800800,cobolli-lestienne,False,20240201
4,11998447,1.0,4,None,FQAsyUF,1.0,3,1706794200,martinez-coric,False,20240201


In [ ]:
MatchRoundInfo.head()

,match_id,round_id,name,slug,cup_round_type,date
0,11998445,5,Round of 16,round-of-16,8.0,20240201
1,11998446,5,Round of 16,round-of-16,8.0,20240201
2,11998447,5,Round of 16,round-of-16,8.0,20240201
3,11998448,5,Round of 16,round-of-16,8.0,20240201
4,11998449,5,Round of 16,round-of-16,8.0,20240201


In [14]:
MatchTimeInfo.head()

,match_id,period_1,period_2,period_3,period_4,period_5,current_period_start_timestamp,date
0,11974053,NaN,NaN,NaN,None,None,NaN,20240201
1,11974066,NaN,NaN,NaN,None,None,NaN,20240201
2,11998445,3259.0,2639.0,4202.0,None,None,1.706817e+09,20240201
3,11998446,2488.0,2375.0,NaN,None,None,1.706804e+09,20240201
4,11998447,3741.0,1913.0,NaN,None,None,1.706798e+09,20240201


In [15]:
MatchTournamentInfo.head()

,match_id,tournament_id,tournament_name,tournament_slug,tournament_unique_id,tournament_category_name,tournament_category_slug,user_count,ground_type,tennis_points,has_event_player_statistics,crowd_sourcing_enabled,has_performance_graph_feature,display_inverse_home_away_teams,priority,competition_type,date
0,11974053,70826,Qualifiers,qualifiers,None,Davis Cup,davis-cup,6909,None,NaN,False,False,False,False,0,2.0,20240201
1,11974066,70826,Qualifiers,qualifiers,None,Davis Cup,davis-cup,6909,None,NaN,False,False,False,False,0,2.0,20240201
2,11998445,126168,"Montpellier, France",montpellier-france,None,ATP,atp,2155,Hardcourt indoor,250.0,False,False,False,False,0,2.0,20240201
3,11998446,126168,"Montpellier, France",montpellier-france,None,ATP,atp,2155,Hardcourt indoor,250.0,False,False,False,False,0,2.0,20240201
4,11998447,126168,"Montpellier, France",montpellier-france,None,ATP,atp,2155,Hardcourt indoor,250.0,False,False,False,False,0,2.0,20240201


In [16]:
MatchVenueInfo.head()

,match_id,city,stadium,venue_id,country,date
0,11974053,Groningen,Martini Plaza,7324,Netherlands,20240201
1,11974066,Vilnius,SEB Arena,36345,Lithuania,20240201
2,11998445,Montpellier,Court Patrice Dominguez,20333,France,20240201
3,11998446,Montpellier,Court Patrice Dominguez,20333,France,20240201
4,11998447,Montpellier,Court Patrice Dominguez,20333,France,20240201


In [17]:
MatchSeasonInfo.head()

,match_id,season_id,name,year,date
0,11974053,55584,Davis Cup 2024,2024,20240201
1,11974066,55584,Davis Cup 2024,2024,20240201
2,11998445,55448,"ATP Montpellier, France Men Singles 2024",2024,20240201
3,11998446,55448,"ATP Montpellier, France Men Singles 2024",2024,20240201
4,11998447,55448,"ATP Montpellier, France Men Singles 2024",2024,20240201


In [18]:
statistics.head()

,match_id,period,statistic_category_name,statistic_name,home_stat,away_stat,compare_code,statistic_type,value_type,home_value,away_value,home_total,away_total,date
0,11998445,ALL,service,aces,12,6,1,positive,event,12,6,NaN,NaN,20240201
1,11998445,ALL,service,double_faults,2,7,2,negative,event,2,7,NaN,NaN,20240201
2,11998445,ALL,service,first_serve,57/101 (56%),53/90 (59%),2,positive,team,57,53,101.0,90.0,20240201
3,11998445,ALL,service,second_serve,42/44 (95%),30/37 (81%),1,positive,team,42,30,44.0,37.0,20240201
4,11998445,ALL,service,first_serve_points,42/57 (74%),39/53 (74%),1,positive,team,42,39,57.0,53.0,20240201


In [19]:
odds.head()

,match_id,market_id,market_name,is_live,suspended,initial_fractional_value,fractional_value,choice_name,choice_source_id,winnig,change,date
0,11974053,1,full_time,False,False,1/4,1/5,1,1994690951,True,-1,20240201
1,11974053,1,full_time,False,False,11/4,10/3,2,1994691017,False,1,20240201
2,11974066,1,full_time,False,False,14/1,14/1,1,1990515041,False,0,20240201
3,11974066,1,full_time,False,False,1/33,3/100,2,1990515081,True,0,20240201
4,11998445,1,full_time,False,False,1/1,4/5,1,1985921173,False,-1,20240201


In [20]:
point_by_point.head()

,match_id,set_id,game_id,point_id,home_point,away_point,point_description,home_point_type,away_point_type,home_score,away_score,serving,scoring,date
0,11998445,3,13,0,1,0,0,6,5,6,7,1,2,20240201
1,11998445,3,13,1,1,1,0,5,6,6,7,1,2,20240201
2,11998445,3,13,2,1,2,0,5,6,6,7,1,2,20240201
3,11998445,3,13,3,1,3,0,5,1,6,7,1,2,20240201
4,11998445,3,13,4,1,4,0,5,1,6,7,1,2,20240201


In [21]:
votes.head()

,match_id,home_vote,away_vote,date
0,11974053,272,61,20240201
1,11974066,124,579,20240201
2,11998445,4740,4091,20240201
3,11998446,897,4558,20240201
4,11998447,4456,780,20240201


In [22]:
tennis_power.head()

,match_id,set_num,game_num,value,break_occurred,date
0,11998445,1,1,-52.80,False,20240201
1,11998445,1,2,48.14,False,20240201
2,11998445,1,3,-51.62,False,20240201
3,11998445,1,4,10.00,False,20240201
4,11998445,1,5,26.60,True,20240201


### 2. Missing Values Analysis

In [23]:
for name,df in datasets.items():
    print("\n",name)
    print(df.isna().sum().sort_values(ascending=False))


 MatchAwayScoreInfo
period_4              35053
period_5              35053
period_4_tie_break    35053
period_5_tie_break    35053
normal_time           35053
period_3_tie_break    34403
period_2_tie_break    32668
period_1_tie_break    32509
period_3              26246
period_2               3346
period_1               3026
current_score          3013
display_score          3013
match_id                  0
date                      0
dtype: int64

 MatchHomeScoreInfo
period_4              35164
period_5              35164
period_4_tie_break    35164
period_5_tie_break    35164
normal_time           35164
period_3_tie_break    34513
period_2_tie_break    32771
period_1_tie_break    32610
period_3              26322
period_2               3357
period_1               3026
current_score          3014
display_score          3014
match_id                  0
date                      0
dtype: int64

 MatchEventInfo
away_team_seed          25170
home_team_seed          25009
first_to_serve 

In [24]:
print("MatchHomeScoreInfoMatchHomeScoreInfo shape:", MatchHomeScoreInfo.shape)
print("MatchAwayScoreInfo shape:", MatchAwayScoreInfo.shape)


# Columns with missing values to investigate
score_columns = [
    "period_4",
    "period_5",
    "period_4_tie_break",
    "period_5_tie_break",
    "normal_time"
]


# Missing value summary for MatchHomeScoreInfo
print("\nMissing values in MatchHomeScoreInfo:")
home_missing = pd.DataFrame({
    "Missing Count": MatchHomeScoreInfo[score_columns].isnull().sum(),
    "Missing Percentage": MatchHomeScoreInfo[score_columns].isnull().mean() * 100
})

print(home_missing)


# Missing value summary for MatchAwayScoreInfo
print("\nMissing values in MatchAwayScoreInfo:")
away_missing = pd.DataFrame({
    "Missing Count": MatchAwayScoreInfo[score_columns].isnull().sum(),
    "Missing Percentage": MatchAwayScoreInfo[score_columns].isnull().mean() * 100
})

print(away_missing)

MatchHomeScoreInfoMatchHomeScoreInfo shape: (35164, 15)
MatchAwayScoreInfo shape: (35053, 15)

Missing values in MatchHomeScoreInfo:
                    Missing Count  Missing Percentage
period_4                    35164               100.0
period_5                    35164               100.0
period_4_tie_break          35164               100.0
period_5_tie_break          35164               100.0
normal_time                 35164               100.0

Missing values in MatchAwayScoreInfo:
                    Missing Count  Missing Percentage
period_4                    35053               100.0
period_5                    35053               100.0
period_4_tie_break          35053               100.0
period_5_tie_break          35053               100.0
normal_time                 35053               100.0


## Missing Values in Match Score Information

The `MatchHomeScoreInfo` and `MatchAwayScoreInfo` DataFrames contain 35,053 rows and 15 columns.

The missing value analysis showed that the following columns contain missing values in all records (100% missing):

- `period_4`
- `period_5`
- `period_4_tie_break`
- `period_5_tie_break`
- `normal_time`

These columns do not provide any information in the current dataset because all values are missing. Therefore, they can be removed during the data cleaning process.

The missing values in `period_4` and `period_5` are expected because most tennis matches follow a best-of-three sets format, meaning that fourth and fifth sets usually do not occur.

In [25]:
statistics_missing = statistics[
    statistics["home_total"].isna() |
    statistics["away_total"].isna()
]

statistics_missing.head()

,match_id,period,statistic_category_name,statistic_name,home_stat,away_stat,compare_code,statistic_type,value_type,home_value,away_value,home_total,away_total,date
0,11998445,ALL,service,aces,12,6,1,positive,event,12,6,NaN,NaN,20240201
1,11998445,ALL,service,double_faults,2,7,2,negative,event,2,7,NaN,NaN,20240201
6,11998445,ALL,service,service_games_played,16,16,3,positive,event,16,16,NaN,NaN,20240201
8,11998445,ALL,points,total,95,96,2,positive,event,95,96,NaN,NaN,20240201
9,11998445,ALL,points,service_points_won,60,55,1,positive,event,60,55,NaN,NaN,20240201


In [26]:
statistics_complete = statistics[
    statistics["home_total"].notna() |
    statistics["away_total"].notna()
]

statistics_complete.head()

,match_id,period,statistic_category_name,statistic_name,home_stat,away_stat,compare_code,statistic_type,value_type,home_value,away_value,home_total,away_total,date
2,11998445,ALL,service,first_serve,57/101 (56%),53/90 (59%),2,positive,team,57,53,101.0,90.0,20240201
3,11998445,ALL,service,second_serve,42/44 (95%),30/37 (81%),1,positive,team,42,30,44.0,37.0,20240201
4,11998445,ALL,service,first_serve_points,42/57 (74%),39/53 (74%),1,positive,team,42,39,57.0,53.0,20240201
5,11998445,ALL,service,second_serve_points,18/44 (41%),16/37 (43%),2,positive,team,18,16,44.0,37.0,20240201
7,11998445,ALL,service,break_points_saved,6/10 (60%),1/6 (16%),1,positive,team,6,1,10.0,6.0,20240201


In [27]:
print("Original:", statistics.shape)
print("Complete:", statistics_complete.shape)
print("Missing:", statistics_missing.shape)

Original: (1358234, 14)
Complete: (533495, 14)
Missing: (824739, 14)


## Missing Values

After checking the missing values in all DataFrames, the following observations were found:

- The `match` DataFrame contains a large number of missing values across several columns.
- In the `odds` DataFrame, the `winning` column contains missing values.
- In the `statistics` DataFrame, the `home_total` and `away_total` columns contain missing values.

These missing values need to be investigated before deciding whether to remove them, replace them, or keep them.

### 3. Duplicate Records Analysis

In [28]:
for name,df in datasets.items():
    print(name,df.duplicated().sum())

MatchAwayScoreInfo 0
MatchHomeScoreInfo 0
MatchEventInfo 0
MatchAwayTeamInfo 0
MatchHomeTeamInfo 0
MatchRoundInfo 0
MatchSeasonInfo 0
MatchTimeInfo 0
MatchTournamentInfo 0
MatchVenueInfo 0
odds 0
point_by_point 0
statistics 0
tennis_power 0
votes 0


Therefore, no exact duplicate rows need to be removed at this stage.





1. **Understanding Data Structure**
   - Examined the number of rows and columns in each DataFrame.
   - Reviewed column names and data types.
   - Investigated the meaning and role of different variables to understand the structure of the tennis data.

2. **Checking Missing Values**
   - Analyzed missing values in each DataFrame.
   - Identified columns with missing data and calculated the number and percentage of missing values.
   - Investigated whether missing values were expected based on the nature of tennis matches or whether they required cleaning.

3. **Detecting Duplicate Records**
   - Checked for exact duplicate rows in each dataset.
   - Duplicate records were identified using the `duplicated()` function.
   - The results were evaluated to determine whether duplicate removal was necessary.

These initial investigations help ensure that the datasets are correctly understood before applying any data preprocessing and analysis methods.